crates/bevy_pbr/src/bloom/mod.rs

```diff
@@ -0,0 +1,760 @@
+            Some((entity, (BloomUniformIndex(index))))
+        })
+        .collect::<Vec<_>>();
+    commands.insert_or_spawn_batch(entities);
+
+    bloom_uniforms
+        .uniforms
+        .write_buffer(&render_device, &render_queue);
+}
+
+#[derive(Component)]
+struct BloomBindGroups {
+    pre_filter_bind_group: BindGroup,
+    down_sampling_bind_groups: Box<[BindGroup]>,
+    up_sampling_bind_groups: Box<[BindGroup]>,
+    up_sampling_final_bind_group: BindGroup,
+}
+
+fn queue_bloom_bind_groups(
+    mut commands: Commands,
+    render_device: Res<RenderDevice>,
+    pipelines: Res<BloomPipelines>,
+    uniforms: Res<BloomUniforms>,
+    views: Query<(Entity, &ViewTarget, &BloomTextures)>,
+) {
+    if let Some(uniforms) = uniforms.uniforms.binding() {
+        for (entity, view_target, textures) in &views {
+            let pre_filter_bind_group = render_device.create_bind_group(&BindGroupDescriptor {
+                label: Some("bloom_pre_filter_bind_group"),
+                layout: &pipelines.down_sampling_bind_group_layout,
+                entries: &[
+                    BindGroupEntry {
+                        binding: 0,
+                        resource: BindingResource::TextureView(view_target.main_texture()),
+                    },
+                    BindGroupEntry {
+                        binding: 1,
+                        resource: BindingResource::Sampler(&pipelines.sampler),
+                    },
+                    BindGroupEntry {
+                        binding: 2,
+                        resource: uniforms.clone(),
+                    },
+                ],
+            });
+
+            let mut down_sampling_bind_groups = Vec::new();
+            for mip in 1..textures.mip_count {
+                let bind_group = render_device.create_bind_group(&BindGroupDescriptor {
+                    label: Some("bloom_down_sampling_bind_group"),
+                    layout: &pipelines.down_sampling_bind_group_layout,
+                    entries: &[
+                        BindGroupEntry {
+                            binding: 0,
+                            resource: BindingResource::TextureView(&BloomTextures::texture_view(
+                                &textures.texture_a,
+                                mip - 1,
+                            )),
+                        },
+                        BindGroupEntry {
+                            binding: 1,
+                            resource: BindingResource::Sampler(&pipelines.sampler),
+                        },
+                        BindGroupEntry {
+                            binding: 2,
+                            resource: uniforms.clone(),
+                        },
+                    ],
+                });
+
+                down_sampling_bind_groups.push(bind_group);
+            }
+
+            let mut up_sampling_bind_groups = Vec::new(); // TODO: Make these boxed slices
```

Comment:
Why the todo? It could just be `Vec::with_capacity()`?

Reply:
It could, but it's slightly more expensive (1 extra usize, iirc) to hold a Vec when we know the capacity won't change.

Reply:
Yep, good point.